### ISSUE TITLE
Suggest borrow when pattern ergonomics cause mismatched lifetimes

### ISSUE THREAD

**Issue by estebank** [rust-lang/rust#69451]
I encountered the following diagnostic:

```
error: lifetime may not live long enough
   --> src/librustc_typeck/check/method/suggest.rs:667:57
    |
667 |                     bound_list.dedup_by_key(|(_, path)| path); // #35677
    |                                              ---------- ^^^^ returning this value requires that `'1` must outlive `'2`
    |                                              |        |
    |                                              |        return type of closure is &'2 mut std::string::String
    |                                              has type `&'1 mut (usize, std::string::String)`
```

The reason it failed is because it is relying on pattern ergonomics to autoborrow the tuple argument, but because of this it fails to unify the lifetimes as a normal closure would. The compiler should detect this case and suggest the appropriate syntax.
